include/crm/common/output.h

```diff
@@ -0,0 +1,482 @@
+     *       such a way that it can be called repeatedly on a previously finished
+     *       object without crashing.
+     *
+     * \param[in,out] out         The output functions structure.
+     * \param[in]     exit_status The exit value of the whole program.
+     */
+    void (*finish) (pcmk__output_t *out, crm_exit_t exit_status);
+
+    /*!
+     * \internal
+     * \brief Finalize output and then immediately set back up to start a new set
+     *        of output.
+     *
+     * This is conceptually the same as calling finish and then init, though in
+     * practice more be happening behind the scenes.
+     *
+     * \note This function differs from finish in that no exit_status is added.
+     *       The idea is that the program is not shutting down, so there is not
+     *       yet a final exit code.  Call finish on the last time through if this
+     *       is needed.
+     *
+     * \param[in,out] out The output functions structure.
+     */
+    void (*reset) (pcmk__output_t *out);
+
+    /*!
+     * \internal
+     * \brief Register a custom message.
+     *
+     * \param[in,out] out        The output functions structure.
+     * \param[in]     message_id The name of the message to register.  This name
+     *                           will be used as the message_id parameter to the
+     *                           message function in order to call the custom
+     *                           format function.
+     * \param[in]     fn         The custom format function to call for message_id.
+     */
+    void (*register_message) (pcmk__output_t *out, const char *message_id,
+                              pcmk__message_fn_t fn);
+
+    /*!
+     * \internal
+     * \brief Call a previously registered custom message.
+     *
+     * \param[in,out] out        The output functions structure.
+     * \param[in]     message_id The name of the message to call.  This name must
+     *                           be the same as the message_id parameter of some
+     *                           previous call to register_message.
+     * \param[in] ...            Arguments to be passed to the registered function.
+     *
+     * \return true if a function was registered for the message and that function
+     *         was called.  false otherwise.
+     */
+    bool (*message) (pcmk__output_t *out, const char *message_id, ...);
+
+    /*!
+     * \internal
+     * \brief Format the output of executed command.
+     *
+     * \param[in,out] out         The output functions structure.
+     * \param[in]     exit_status The exit value of the executed command.
+     * \param[in]     src         The source of this output - stdout or stderr.
+     * \param[in]     buf         The raw command output to be formatted.
+     */
+    void (*command_output) (pcmk__output_t *out, int exit_status,
+                            pcmk__cmd_output_source src, const char *buf);
```

Comment:
Is `buf` passed into `command_output` meant to be collected over the
previous invocations of `out->method`, since the CLI tool invoking this
is effectively to refer to itself, and previously set how this output is
to be collected in the first place (through a formatter)?  If so, then
`buf` is redundant, since the "object" behind output will have this `buf`
stored internally and will want to refer to it at the point of
`command_output` (presumably just before exiting, so it could actually
be codependent with the destructors), too?

Also opens questions whether this accumulate to heap, output at one
go at the end is a preferred model to optimize API for.

Reply:
At the moment, `buf` is not collected by the formatted output API at all.  It's up to the caller to collect the output and then just pass it to `command_output`.  See the validate function in tools/stonith_admin.c.  I'm just replacing a portion of what was already going on in there.

Reply:
The confusion is inherent to the purpose: this isn't output of the command calling the function, it's output of a command that that command ran. I.e. in this case, stonith_admin runs a fence agent, then displays that agent's stdout and stderr.

Reply:
I'd be happy to go with a different, shorter, name here if anyone has a good idea.

Reply:
Maybe "subcommand_output" or "subprocess_output" would make it clearer

It allows a command to output the output of a command :)

Reply:
Pardon my ignorance, from an earlier discussion, I seem to recall that
the data-structured output of CLI tools would be rather self-contained,
incl. the exit status and what not.  Because only in that case it makes
sense to select exactly one of `stdout` and `stderr` source indication,
(even if some moments later, it seems that exit status would suffice
fully).

Involving 3rd party executable here doesn't seem very compatible, since
there are cases (next to all?) you want to pass on both sinks, and this
interface will either cause unnecessary data redundancy (duplicate info
about exit statuses) and churn at the acceptor of the structured data,
or loss of some messaging possibly important in the context.

Reply:
This is the exit status of the executed subprocess, not the exit status of the caller of this function. So by the time stdout/stderr is received, the exit status is known. (This is not piped partial output from a still-running subprocess, but complete output from a finished subprocess.)

Text mode ignores the exit status; XML includes it as a code="<N>" attribute, which I wouldn't consider significant redundancy to list twice. But I could see using an interface like 

    void (*subprocess_output) (pcmk__output_t *out, int exit_status,
                                                const char *proc_stdout, const char *proc_stderr);

where either could be NULL to skip outputting it. That would avoid the need for the enum as well.

Reply:
Yeah, this was the other way I had been leaning in the beginning.  I thought I might come up with some other potential sources to go into that enum and make it look a little less overthought.  Nothing's come up, so the enum and multiple calls seems a little pointless.  Converting to the single call style brings up two points:

* On the text output, I'm not sure that I can do anything besides just dump stdout first and then stdout second, which could potentially looks confusing depending on what subprocess is being run.  Imagine compiler output where the regular output and error messages are all separate.  Don't know that I can do much of anything.

* On the XML output, I'm going to have to put the code="" on each block, as you mentioned.  I was already doing this, but one option could be to wrap the stdout node and the stderr node in an additional layer, and that layer would have the code attribute.

Reply:
Agreed about the text output, the good thing about being internal API is we can easily change it later if we need different behavior. For the XML, the extra layer makes sense.

Reply:
I vote for the extra layer, so the xml structure would be like this both for xml and plaintext:
```
<command-output status="">
  <stdout></stdout>
  <stderr></stderr>
</command-output>
```
Having status at two places does not make sense to me. It allows each of the two places to have different status value which is just bad.

Also following syntax allows several stdouts and stderrs to be listed which seems like a mistake - what would be meaning of that?
```
<define name="element-validation">
  <element name="validation">
    ...
    <zeroOrMore>
      <ref name="command-output" />
    </zeroOrMore>
  </element>
</define>
<define name="command-output">
  <element name="command-output">
    <attribute name="source">
      <choice>
        <value>stdout</value>
        <value>stderr</value>
      </choice>
    </attribute>
    <text />
  </element>
</define>
```